Different behaviour for code that sets the crate type after switching `build` && `test` command

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9f365edcfc4695eefcbcf6bd661cea52)

When you first time to click `test` in top left corner, the error message is:

```
   Compiling playground v0.0.1 (/playground)
    Finished test [unoptimized + debuginfo] target(s) in 0.55s
     Running unittests src/lib.rs (target/debug/deps/playground-e330e1c88dfd4fd8)
   Doc-tests playground
error: doctest failed, to rerun pass `--doc`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

running 1 test
test src/lib.rs - demo (line 6) ... FAILED

failures:

---- src/lib.rs - demo (line 6) stdout ----
error[E0412]: cannot find type `aaa` in this scope
 --> src/lib.rs:9:10
  |
5 | playground::demo!();
  | ------------------- similarly named struct `Aaa` defined here
6 | type T = aaa; // FIXME
  |          ^^^ help: a struct with a similar name exists: `Aaa`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0412`.
Couldn't compile the test.

failures:
    src/lib.rs - demo (line 6)

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.16s
```

Then I click `build`(or `run`) to execute command, waiting them finished, and then click `test`, the error message is: 

```
  Compiling playground v0.0.1 (/playground)
error: the `#[proc_macro]` attribute is only usable with crates of the `proc-macro` crate type
  --> src/lib.rs:10:1
   |
10 | #[proc_macro]
   | ^^^^^^^^^^^^^

error[E0432]: unresolved import `proc_macro`
 --> src/lib.rs:3:5
  |
3 | use proc_macro::TokenStream;
  |     ^^^^^^^^^^ use of undeclared crate or module `proc_macro`
  |
help: there is a crate or module with a similar name
  |
3 | use proc_macro2::TokenStream;
  |     ~~~~~~~~~~~

For more information about this error, try `rustc --explain E0432`.
error: could not compile `playground` (lib test) due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `playground` (lib) due to 2 previous errors
```

Different error message.

--- a/tests/spec/features/test_action_spec.rb
+++ b/tests/spec/features/test_action_spec.rb
@@ -1,0 +1,32 @@
+require 'spec_helper'
+require 'support/editor'
+require 'support/playground_actions'
+
+RSpec.feature "Testing the code", type: :feature, js: true do
+  include PlaygroundActions
+
+  before { visit '/' }
+
+  scenario "when a crate type is provided" do
+    code = <<~EOF
+      #![crate_type = "proc-macro"]
+
+      use proc_macro::TokenStream;
+
+      #[proc_macro]
+      pub fn demo(input: TokenStream) -> TokenStream {
+          input
+      }
+    EOF
+    editor.set(code)
+
+    in_build_menu { click_on(build_button: "Test") }
+    within(:output, :stdout) do
+      expect(page).to have_content "running 0 tests"
+    end
+  end
+
+  def editor
+    Editor.new(page)
+  end
+end

--- a/ui/frontend/actions.ts
+++ b/ui/frontend/actions.ts
@@ -263,7 +263,11 @@ function performAutoOnly(): ThunkAction {
 
 const performExecuteOnly = (): ThunkAction => performCommonExecute('bin', false);
 const performCompileOnly = (): ThunkAction => performCommonExecute('lib', false);
-const performTestOnly = (): ThunkAction => performCommonExecute('lib', true);
+const performTestOnly = (): ThunkAction => (dispatch, getState) => {
+  const state = getState();
+  const crateType = getCrateType(state);
+  return dispatch(performCommonExecute(crateType, true));
+};
 
 interface GenericApiFailure {
   error: string;

